docs/architecture/grpc-for-wcf-developers/apm.md

```diff
@@ -1,11 +1,221 @@
+
+### Distributed tracing with DiagnosticSource
+
+.NET Core has an internal module that maps very well to distributed traces and spans: [DiagnosticSource](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md). As well as providing a simple way to produce and consume diagnostics within a process, the DiagnosticSource module has the concept of an *Activity*, which is effectively an implementation of a distributed trace, or a span within a trace. The internals of the module take care of parent/child activities, including allocating identifiers.
+
+Because DiagnosticSource is a part of the core framework, it is supported by several core components, including `HttpClient`, Entity Framework Core and ASP.NET Core, including explicit support in the gRPC framework. When ASP.NET Core receives a request, it will check for a pair of HTTP headers matching the [W3C Trace Context](https://www.w3.org/TR/trace-context) standard. If the headers are found, an Activity is started using the identity values and context from the headers. If no headers are found, an Activity is started with generated identity values that match the standard format. Any diagnostics generated by the framework or by application code during the lifetime of this Activity can be tagged with the trace and span identifiers. The `HttpClient` support extends this further by checking for a current Activity on every request and automatically adding the trace headers to the outgoing request.
+
+The gRPC client and server libraries include explicit support for DiagnosticSource and Activity, and will create activities and apply and use header information automatically.
+
+> [!NOTE]
+> All of this only happens if a *listener* is consuming the diagnostic information. If there is no listener, no diagnostics are written and no activities are created.
+
+### Implementing your own DiagnosticSource
+
+Although a good quantity of data is automatically generated by ASP.NET Core, including gRPC, as well as Entity Framework Core and `HttpClient`, you may wish to add your own diagnostics or create explicit spans within your application code. You can create a named `DiagnosticSource` and write event data to it, or create an `Activity` to record a specific operation. `DiagnosticSource` is an abstract class; `DiagnosticListener` inherits from it and can be used to create an instance. The class is intended to be used as a singleton; it is common to create it as a static field in an internal class.
```

Comment:
Maybe link to the Activity guide in CoreFx? https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md#activity-user-guide